src/webapi/tests/integ.rs

```diff
@@ -0,0 +1,60 @@
+use once_cell::sync::Lazy;
+use std::{net::TcpListener, str::FromStr};
+use tracing::Level;
+use webapi::{
+    startup::run,
+    telemetry::{get_subscriber, init_subscriber},
+};
+
+// Ensure that the `tracing` stack is only initialised once using `once_cell`
+static TRACING: Lazy<()> = Lazy::new(|| {
+    let subscriber_name = "test".to_string();
+    if let Ok(env_var) = std::env::var("TEST_LOG") {
+        let level = tracing::Level::from_str(&env_var)
+            .unwrap_or(Level::INFO)
+            .as_str();
+        let subscriber = get_subscriber(subscriber_name, level.to_owned(), std::io::stdout);
+        init_subscriber(subscriber);
+    } else {
+        let subscriber = get_subscriber(subscriber_name, "info".to_string(), std::io::sink);
+        init_subscriber(subscriber);
+    };
+});
+
+pub struct TestApp {
+    pub address: String,
+}
+
+async fn spawn_app() -> TestApp {
+    // The first time `initialize` is invoked the code in `TRACING` is executed.
+    // All other invocations will instead skip execution.
+    Lazy::force(&TRACING);
+
+    let listener = TcpListener::bind("127.0.0.1:0").expect("Failed to bind random port");
+    // We retrieve the port assigned to us by the OS
+    let port = listener.local_addr().unwrap().port();
+    let address = format!("http://127.0.0.1:{}", port);
+
+    let server = run(listener).expect("Failed to bind address");
+    let _ = tokio::spawn(server);
+    TestApp { address }
+}
+
+#[tokio::test]
+async fn ping_works() {
+    // Arrange
+    let app = spawn_app().await;
+    let client = reqwest::Client::new();
+
+    // Act
+    let response = client
+        // Use the returned application address
+        .get(&format!("{}/ping", &app.address))
+        .send()
+        .await
+        .expect("Failed to execute request.");
+
+    // Assert
+    assert!(response.status().is_success());
+    assert_eq!(response.bytes().await.unwrap(), "\"pong\"".to_string());
```

Comment:
Since we're returning JSON, I think it would be cleaner to check this as a JSON, not as a string.

Reply:
Changed